overlord/snapstate/refresh.go

```diff
@@ -0,0 +1,217 @@
+// currently running.
+//
+// Since services are stopped during the update this provides a good early
+// precondition check. The check is also deliberately racy as existing snap
+// commands can fork new processes or existing processes can die. After the
+// soft check passes the user is free to start snap applications and block the
+// hard check.
+func SoftNothingRunningRefreshCheck(info *snap.Info) error {
+	return genericRefreshCheck(info, func(app *snap.AppInfo) bool {
+		return app.IsService()
+	})
+}
+
+// HardNothingRunningRefreshCheck looks if there are any undesired processes alive.
+//
+// The check is designed to run late in the refresh pipeline, after stopping
+// snap services. At this point non-enduring services should be stopped, hooks
+// should no longer run, and applications should be barred from running
+// externally (e.g. by using a new inhibition mechanism for snap run).
+//
+// The check fails if any process belonging to the snap, apart from services
+// that are enduring refresh, is still alive. If a snap is busy it cannot be
+// refreshed and the refresh process is aborted.
+func HardNothingRunningRefreshCheck(info *snap.Info) error {
+	return genericRefreshCheck(info, func(app *snap.AppInfo) bool {
+		// TODO: use a constant instead of "endure"
+		return app.IsService() && app.RefreshMode == "endure"
+	})
+}
+
+// BusySnapError indicates that snap has apps or hooks running and cannot refresh.
+type BusySnapError struct {
+	snapName      string
+	pids          []int
+	busyAppNames  []string
+	busyHookNames []string
+}
+
+// Error formats an error string describing what is running.
+func (err *BusySnapError) Error() string {
+	switch {
+	case len(err.busyAppNames) > 0 && len(err.busyHookNames) > 0:
+		return fmt.Sprintf("snap %q has running apps (%s) and hooks (%s)",
+			err.snapName, strings.Join(err.busyAppNames, ", "), strings.Join(err.busyHookNames, ", "))
+	case len(err.busyAppNames) > 0:
+		return fmt.Sprintf("snap %q has running apps (%s)",
+			err.snapName, strings.Join(err.busyAppNames, ", "))
+	case len(err.busyHookNames) > 0:
+		return fmt.Sprintf("snap %q has running hooks (%s)",
+			err.snapName, strings.Join(err.busyHookNames, ", "))
+	default:
+		return fmt.Sprintf("snap %q has running apps or hooks", err.snapName)
+	}
+}
+
+// Pids returns the set of process identifiers that are running.
+//
+// Since this list is a snapshot it should be only acted upon if there is an
+// external synchronization system applied (e.g. all processes are frozen) at
+// the time the snapshot was taken.
+//
+// The list is intended for snapd to forcefully kill all processes for a forced
+// refresh scenario.
+func (err BusySnapError) Pids() []int {
+	return err.pids
+}
+
+// parsePid parses a string as a process identifier.
+func parsePid(text string) (int, error) {
+	pid, err := strconv.Atoi(text)
+	if err == nil && pid <= 0 {
+		return 0, fmt.Errorf("cannot parse pid %q", text)
+	}
```

Comment:
It would be highly unusual if pid == 1 too and I'm a little concerned if there is a bug that we might forcibly kill pid 1 and nuke the system. That said, considering snapd running in containers, pid 1 might not be as interesting in the container. I bring this up not to change anything but to simply think through if this is something we'd want to explicitly avoid here.